p2p/simulations/events_test.go

```diff
@@ -0,0 +1,170 @@
+package simulations
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"testing"
+	"time"
+
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/p2p/discover"
+	"github.com/ethereum/go-ethereum/p2p/simulations/adapters"
+)
+
+func init() {
+	log.Root().SetHandler(log.LvlFilterHandler(log.LvlInfo, log.StreamHandler(os.Stderr, log.TerminalFormat(true))))
+}
+
+func TestNodeUpAndConn(t *testing.T) {
+
+	var quitC []chan struct{}
+	defer func() {
+		for _, q := range quitC {
+			if q != nil {
+				close(q)
+			}
+		}
+	}()
+
+	// create simulation network with 20 testService nodes
+	adapter := adapters.NewSimAdapter(adapters.Services{
+		"test": newTestService,
+	})
+	network := NewNetwork(adapter, &NetworkConfig{
+		DefaultService: "test",
+	})
+	nodeCount := 3
+	ids := make([]discover.NodeID, nodeCount)
+
+	for i := 0; i < nodeCount; i++ {
+		conf := adapters.RandomNodeConfig()
+		node, err := network.NewNodeWithConfig(conf)
+		ids[i] = node.ID()
+		if err != nil {
+			t.Fatalf("error creating node: %s", err)
+		}
+	}
+
+	trigger := make(chan discover.NodeID)
+	events := make(chan *Event)
+	sub := network.Events().Subscribe(events)
+	defer sub.Unsubscribe()
+	quitC = append(quitC, make(chan struct{}))
+
+	action := func(ctx context.Context) error {
+		go func(quitC chan struct{}) {
+			for {
+				select {
+				case ev := <-events:
+					if ev == nil {
+						panic("got nil event")
+					}
+					if ev.Type == EventTypeNode {
+						if ev.Node.Up {
+							log.Info("got node up event", "event", ev, "node", ev.Node.Config.ID)
+							trigger <- ev.Node.Config.ID
+						}
```

Comment:
is actually ckecking for `ev.Node.Up == false` useful? We would only expect node up events at this point right?

Reply:
I'm paranoid about the snapshots showing nodes as down, where I don't know the cause :) but good point.